### PR TITLE
Kixer ads: Fire ad viewability when ad is in viewport

### PR DESCRIPTION
**Pull request by jvelis**
We would like to improve viewability numbers by using the position in the viewport to fire the view.

Using window.context.observeIntersection API to detect the position.

### CODE REVIEW

**Comment by jvelis** [comment]
@ampprojectbot retry!

File: ads/kixer.js [hunk 1]
```diff
@@ -39,11 +44,34 @@ export function kixer(global, data) {
     d.removeEventListener('load', kxload, false);
     if (d.childNodes.length > 0) {
       global.context.renderStart();
+      view_interval = setInterval(function() {
+        kxview_check(); // Once the ad has loaded, start checking for an ad view
```
Comment (lannka) [inline ads/kixer.js]: is such a poll necessary? `observeIntersection` will call the callback when there is a change.

Comment (jvelis) [inline ads/kixer.js]: Hi Lanka! Thanks for your comment. It makes sense, however, this helped to make sure the ad was in the viewport for at least 1 second and prevented duplicate/early calls to the callback function.

Comment (lannka) [inline ads/kixer.js]: if all you want is "visible for 1s", then isn't `setTimeout` more appropriate than `setInterval`?

File: ads/kixer.js [hunk 2]
```diff
@@ -43,7 +48,33 @@ export function kixer(global, data) {
       global.context.noContentAvailable();
     }
   };
-  d.addEventListener('load', kxload, false);
+  d.addEventListener('load', kxload, false); // Listen for the kixer load event
+
+  const kxview_check = function(coords) {
+    if (coords.intersectionRect.height > coords.boundingClientRect.height / 2) {
+      if (viewed === false && view_timer == null) {
+        view_timer = setTimeout(function() {
```
Comment (jvelis) [inline ads/kixer.js]: Replaced setInterval for setTimeout as suggested by Lannka.

File: ads/kixer.js [hunk 3]
```diff
@@ -43,7 +48,33 @@ export function kixer(global, data) {
       global.context.noContentAvailable();
     }
   };
-  d.addEventListener('load', kxload, false);
+  d.addEventListener('load', kxload, false); // Listen for the kixer load event
+
+  const kxview_check = function(coords) {
```
Comment (lannka) [inline ads/kixer.js]: please use camel case for all local vars, e.g. `kxviewCheck`.

File: ads/kixer.js [hunk 4]
```diff
@@ -43,7 +48,33 @@ export function kixer(global, data) {
...
+          view_timer = null;
+          if (in_view === true) {
+            if (typeof __kx_viewability.process_locked === 'function') {
+              viewed = true;
+              __kx_viewability.process_locked(data.adslot);
+            }
+          }
+        }, 900);
+      }
+      in_view = true;
```
Comment (lannka) [inline ads/kixer.js]: nit: you can move this out of the if block.
```
inView = coords.intersectionRect.height > coords.boundingClientRect.height / 2;
```

File: ads/kixer.js [hunk 5]
```diff
@@ -43,7 +48,33 @@ export function kixer(global, data) {
       global.context.noContentAvailable();
     }
   };
-  d.addEventListener('load', kxload, false);
+  d.addEventListener('load', kxload, false); // Listen for the kixer load event
+
+  const kxview_check = function(coords) {
+    if (coords.intersectionRect.height > coords.boundingClientRect.height / 2) {
```
Comment (lannka) [inline ads/kixer.js]: can you use
`coords.intersectionRatio > 0.5`?

File: ads/kixer.js (hunk 3 above)
Comment (lannka) [inline ads/kixer.js]: nit: `coords` -> `intersectionEntry`

File: ads/kixer.js (hunk 4 above)
Comment (jvelis) [inline ads/kixer.js]: Thanks Lannka. This should be out of the block.

**Comment by jvelis** [comment]
@ampprojectbot retry!

**Comment by jvelis** [comment]
/to lannka zhouyx Hi Lannka, the changes are ready. Please let me know what you think.

**Comment by jvelis** [comment]
Hi @zhouyx and @lannka , my changes are ready. I would appreciate if someone could take a look at my PR whenever possible. Thanks for your help guys!

File: ads/kixer.js [hunk 6]
```diff
@@ -43,7 +48,31 @@ export function kixer(global, data) {
...
+          if (inView === true) {
+            if (typeof __kx_viewability.process_locked === 'function') {
+              viewed = true;
+              __kx_viewability.process_locked(data.adslot);
+            }
+          }
+        }, 900);
+      }
+    }
+  };
```
Comment (lannka) [inline ads/kixer.js]: if you want it to be visible for 900ms before firing, should you reset the timer when `inView === false`?

File: ads/kixer.js [hunk 7]
```diff
@@ -43,7 +48,31 @@ export function kixer(global, data) {
       global.context.noContentAvailable();
     }
   };
-  d.addEventListener('load', kxload, false);
+  d.addEventListener('load', kxload, false); // Listen for the kixer load event
+
+  const kxviewCheck = function(intersectionEntry) {
+    inView = intersectionEntry.intersectionRatio > 0.5; // Half of the unit is in the viewport
+    if (inView === true) {
```
Comment (lannka) [inline ads/kixer.js]: you can just do `if (inView)`

File: ads/kixer.js [hunk 8]
```diff
@@ -43,7 +48,31 @@ export function kixer(global, data) {
...
+      if (viewed === false && viewTimer == null) {
+        viewTimer = setTimeout(function() {
```
Comment (lannka) [inline ads/kixer.js]: add a comment what the timer is for.

File: ads/kixer.js [hunk 9]
```diff
@@ -43,7 +48,31 @@ export function kixer(global, data) {
...
+          clearTimeout(viewTimer);
+          viewTimer = null;
+          if (inView === true) {
+            if (typeof __kx_viewability.process_locked === 'function') {
+              viewed = true;
+              __kx_viewability.process_locked(data.adslot);
+            }
+          }
+        }, 900);
```
Comment (lannka) [inline ads/kixer.js]: do you want 900ms? or 1000ms

File: ads/kixer.js (hunk 6 above)
Comment (jvelis) [inline ads/kixer.js]: Good point, although the timer is effectively recreated when kxviewCheck() is called again by observeIntersection.

File: ads/kixer.js (hunk 9 above)
Comment (jvelis) [inline ads/kixer.js]: 900ms would match our current viewability settings.

**Comment by jvelis** [comment]
Hi @lannka  I made the changes you suggested. Please let me know your thoughts. Thanks!

File: ads/kixer.js (hunk 6 above)
Comment (lannka) [inline ads/kixer.js]: this is not using the memory wisely. please clear it manually

**Comment by lannka** [comment]
@jvelis in case in missed the last comment:
https://github.com/ampproject/amphtml/pull/8289#discussion_r113783199

File: ads/kixer.js (hunk 6 above)
Comment (jvelis) [inline ads/kixer.js]: @lannka Thanks for noticing that. The timer should now be cleared when inView is false to use memory better.

**Comment by jvelis** [comment]
@lannka Thanks for the last comments. Per your suggestion, I made some changes to clear the timeout more efficiently.

**Pull request merged**